src/exchange_price_source.ts

```diff
@@ -0,0 +1,130 @@
+import { ExchangeAdapter, Ticker } from './exchange_adapters/base'
+import BigNumber from 'bignumber.js'
+import { PriceSource, WeightedPrice } from './price_source'
+import { Exchange, OracleCurrencyPair } from './utils'
+import { MetricCollector } from './metric_collector'
+import { individualTickerChecks } from './aggregator_functions'
+
+export interface OrientedExchangePair {
+  exchange: Exchange
+  symbol: OracleCurrencyPair
+  to_invert: boolean
+}
+
+export interface ExchangePriceSourceConfig {
+  pairs: OrientedExchangePair[]
+}
+
+export type OrientedAdapter = [ExchangeAdapter, boolean]
+
+export type PairData = {
+  bid: BigNumber
+  ask: BigNumber
+  baseVolume: BigNumber
+  quoteVolume: BigNumber
+}
+
+function invertPair(pair: PairData): PairData {
+  return {
+    bid: pair.ask.exponentiatedBy(-1),
+    ask: pair.bid.exponentiatedBy(-1),
+    baseVolume: pair.quoteVolume,
+    quoteVolume: pair.baseVolume,
+  }
+}
+
+function tickerToPairData(ticker: Ticker): PairData {
+  return {
+    bid: ticker.bid,
+    ask: ticker.ask,
+    baseVolume: ticker.baseVolume,
+    // Using lastPrice to convert from baseVolume to quoteVolume, whereas
+    // ideally one would use the baseVolume execution's VWAP.
+    quoteVolume: ticker.baseVolume.multipliedBy(ticker.lastPrice),
```

Comment:
Not saying this is incorrect, but I'm wondering if there a case to use the midpoint of the bid and ask over the last price?

Reply:
Ideally we'd have the quote volume (quote notional) provided on the ticker. Since we don't, we have to calculate it from the base volume, and to do that correctly we'd need the average execution price for the ticker (previous 24h). Since we don't have it either, we have to choose a price to use in the conversion. I think both mid and last are not ideal, but since the last was actually on one trade it's slightly less wrong (whereas the mid could have no relation at all with the price of executions over the past 24h).

Reply:
Done on [4f5754c4d0fea068999aed30528bd5d0c7661ac3](https://github.com/celo-org/celo-oracle/commit/4f5754c4d0fea068999aed30528bd5d0c7661ac3).